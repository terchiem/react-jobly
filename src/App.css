.App {
}


.debug {
  width: 600px;
  margin: 20px auto;
  padding: 10px 20px;
  border: 1px solid #aaa;
  background-color: #eee;
  color: #361515;
}