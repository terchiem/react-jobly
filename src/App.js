import React, { useState, useEffect } from 'react';
import { BrowserRouter } from 'react-router-dom';
import jwt from 'jsonwebtoken';
import { TOKEN_STORAGE_KEY } from './config';
import JoblyApi from './JoblyApi';
import UserContext from './UserContext';
import useLocalStorage from './hooks/useLocalStorage';
import './App.css';

// components
import NavBar from './NavBar';
import Routes from './Routes';
import LoadingSpinner from './LoadingSpinner';

/** Top level component for Jobly App. */

function App() {

  const [token, setToken, removeToken] = useLocalStorage(TOKEN_STORAGE_KEY);
  const [currentUser, setCurrentUser] = useState(null);
  const [loading, setLoading] = useState(true);

  // Call api to get user data from the token username, and set in state
  useEffect(() => {
    async function fetchUser() {
      try {
        const { username } = jwt.decode(token);
        const user = await JoblyApi.getUser(username);
        setCurrentUser(user);  
      } catch (err) {
        setCurrentUser(null);
      }
      setLoading(false);
    }
    setLoading(true);
    fetchUser()
  }, [token]);


  /** Logs out current user by removing the token from storage */
  function logOut() {
    removeToken();
    setCurrentUser(null);
  }

  /** Updates the currentUser to match the edited user received from api */
  function setEditedUser(editedUser) {
    setCurrentUser((user) => ({
      ...editedUser,
      jobs: user.jobs
    }));
  }

  /** Updates the currentUser's job list after applying for one */
  function updateUserJobs(job) {
    setCurrentUser(user => ({
      ...user,
      jobs: [...user.jobs, job]
    }));
  }

  if (loading) {
    return <LoadingSpinner />;
  }

  return (
    <div className="App">
      <UserContext.Provider value={{ currentUser, token }}>
        <BrowserRouter>
          <NavBar logOut={logOut} />
          <Routes 
            setToken={setToken} 
            removeToken={removeToken} 
            setEditedUser={setEditedUser} 
            updateUserJobs={updateUserJobs}
          />
        </BrowserRouter>
      </UserContext.Provider>
    </div>
  );
}

export default App;
