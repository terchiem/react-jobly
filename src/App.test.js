import React from 'react';
import { render } from '@testing-library/react';
import App from './App';

test('renders without crashing', () => {
  render(<App />)
});

test('matches snapshot', () =>{
  const { asFragment } = render(<App />);
  expect(asFragment()).toMatchSnapshot();
});