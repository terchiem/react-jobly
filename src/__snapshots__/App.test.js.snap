// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <nav
      class="NavBar"
    >
      <a
        href="/"
        id="NavBar-logo"
      >
        Jobly
      </a>
      <ul>
        <li>
          <a
            href="/login"
          >
            Login
          </a>
        </li>
      </ul>
    </nav>
    <div
      class="Home"
    >
      <h1>
        Jobly
      </h1>
      <p>
        <a
          class="Home-login"
          href="/login"
        >
          Login
        </a>
      </p>
    </div>
  </div>
</DocumentFragment>
`;
